mlir/lib/Mitigation/Transforms/MitigationMethods/Zne.cpp

```diff
@@ -243,7 +187,168 @@ FlatSymbolRefAttr ZneLowering::getOrInsertFoldedCircuit(Location loc, PatternRew
+    // Get the device
+    quantum::DeviceInitOp deviceInitOp = *fnOp.getOps<quantum::DeviceInitOp>().begin();
+
+    StringAttr lib = deviceInitOp.getLibAttr();
+    StringAttr name = deviceInitOp.getNameAttr();
+    StringAttr kwargs = deviceInitOp.getKwargsAttr();
+
+    TypeRange originalTypes = op.getArgs().getTypes();
+    SmallVector<Type> typesFolded(originalTypes.begin(), originalTypes.end());
+    Type indexType = rewriter.getIndexType();
+    typesFolded.push_back(indexType);
+
+    // Function without measurements: Create function without measurements and with qreg as last
+    // argument
+    func::FuncOp fnWithoutMeasurementsOp;
+    func::FuncOp fnWithMeasurementsOp;
+    if (foldingAlgorithm == Folding(1)) {
+        FlatSymbolRefAttr fnWithoutMeasurementsRefAttr =
+            getOrInsertFnWithoutMeasurements(loc, rewriter, op);
+        fnWithoutMeasurementsOp =
+            SymbolTable::lookupNearestSymbolFrom<func::FuncOp>(op, fnWithoutMeasurementsRefAttr);
+        // Function with measurements: Modify the original function to take a quantum register as
+        // last arg and keep measurements
+        FlatSymbolRefAttr fnWithMeasurementsRefAttr =
+            getOrInsertFnWithMeasurements(loc, rewriter, op);
+        fnWithMeasurementsOp =
+            SymbolTable::lookupNearestSymbolFrom<func::FuncOp>(op, fnWithMeasurementsRefAttr);
+    }
+    else {
+    }
+    rewriter.setInsertionPointToStart(moduleOp.getBody());
+
+    FunctionType fnFoldedType = FunctionType::get(ctx, /*inputs=*/
+                                                  typesFolded,
+                                                  /*outputs=*/fnOp.getResultTypes());
+
+    func::FuncOp fnFoldedOp = rewriter.create<func::FuncOp>(loc, fnFoldedName, fnFoldedType);
+    Value c0, c1, numberQubitsValue;
+    fnFoldedOp.setPrivate();
+    if (foldingAlgorithm == Folding(1)) {
+        Block *fnFoldedOpBlock = fnFoldedOp.addEntryBlock();
+        rewriter.setInsertionPointToStart(fnFoldedOpBlock);
+        // Loop control variables
+        c0 = rewriter.create<index::ConstantOp>(loc, 0);
+        c1 = rewriter.create<index::ConstantOp>(loc, 1);
+        TypedAttr numberQubitsAttr = rewriter.getI64IntegerAttr(numberQubits);
+        numberQubitsValue = rewriter.create<arith::ConstantOp>(loc, numberQubitsAttr);
+        rewriter.create<quantum::DeviceInitOp>(loc, lib, name, kwargs);
+    }
+    else {
+        rewriter.cloneRegionBefore(fnOp.getBody(), fnFoldedOp.getBody(), fnFoldedOp.end());
+
+        Block *fnFoldedOpBlock = &fnFoldedOp.getBody().front();
+        rewriter.setInsertionPointToStart(fnFoldedOpBlock);
+        // Loop control variables
+        c0 = rewriter.create<index::ConstantOp>(loc, 0);
+        c1 = rewriter.create<index::ConstantOp>(loc, 1);
+
+        fnFoldedOpBlock->addArgument(fnFoldedOp.getArgumentTypes().front(), loc);
+    }
+
+    if (foldingAlgorithm == Folding(1)) {
```

Comment:
Not sure, can we move all the `foldingAlgorithm == Folding(1)` in a single if statement?

Reply:
Done.